fetch-in-react-practice/starter-files/client/components/app.jsx

```diff
@@ -0,0 +1,144 @@
+    *   - replace the old todos array in state with the new one
+    *
+    * DO NOT MUTATE the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+    const options = {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(newTodo)
+    };
+
+    fetch('/api/todos', options)
+      .then((response) => response.json())
+      .then((data) => {
+        setErr(false);
+        const copyState = todos.slice(0);
+        setTodos(todos.concat(data));
+      })
+      .catch((error) => {
+        console.error(error);
+        setErr(true);
+      });
+  }
+
+  function toggleCompleted(todoId) {
+    /**
+     * Find the index of the todo with the matching todoId in the state array.
+     * Get its "isCompleted" status.
+     * Make a new object containing ONE PROPERTY: the opposite "isCompleted" status.
+     * Use fetch to send a PATCH request to `/api/todos/${todoId}`
+     * Then 😉, once the response JSON is received and parsed,
+     *   - create a shallow copy of the todos array from state
+     *   - replace the old todo with the todo received from the server
+     *   - replace the old todos in the state with the new one (you know the index).
+     *
+     * NOTE: "toggle" means to flip back and forth, so clicking a todo
+     * in the list repeatedly should "toggle" its isCompleted status back and forth.
+     *
+     * DO NOT try to calculate the index of the todo by subtracting 1 from the id.
+     *
+     * DO NOT MUTATE the original state array, nor any objects within it.
+     * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+     *
+     * TIP: Be sure to SERIALIZE the updates in the body with JSON.stringify()
+     * And specify the "Content-Type" header as "application/json"
+     */
+    let oppIsCompleted;
+    let index;
+    let request;
+    for (let i = 0; i < todos.length; i++) {
+      if (todos[i].todoId === todoId) {
+        index = i;
+        oppIsCompleted = !todos[i].isCompleted;
+        request = { isCompleted: oppIsCompleted };
+      }
+    }
+    const options = {
+      method: 'PATCH',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(request)
+    };
+    fetch(`/api/todos/${todoId}`, options)
+      .then((response) => response.json())
+      .then((data) => {
+        setErr(false);
+        data.isCompleted = oppIsCompleted;
+        const copyState = todos.slice(0);
+        copyState[index] = data;
+        setTodos(copyState);
+      })
```

Comment:
Since the `PATCH` may take a while, there's no guarantee the patched todo item will be in the same place you left it.